### ISSUE TITLE
/blog/posts/20140317-RefreshingInstanceStoreHandleInWorkflowFoundation.html

### ISSUE THREAD

**Issue by damirarh** [damirarh/damirscorner-utterances#10]
Imported 

URL: https://www.damirscorner.com/blog/posts/20140317-RefreshingInstanceStoreHandleInWorkflowFoundation.html

**Comment by damirarh** [comment]
Imported comment written by **hamid rezaeirad** on **2015-09-06T12:33:36**

<p>Hi,</p><p>I want to persist the workflow instance and load it.I create a WF app and reference it to a windows application and persist the WF but when I close the application and Open it again I give error :</p><p>The execution of an InstancePersistenceCommand was interrupted because the instance 'ba8ab62d-5642-4ebe-9c9d-bbb9d41e3dbc' is locked by a different instance owner. This error usually occurs because a different host has the instance loaded. The instance owner ID of the owner or host with a lock on the instance is 'ccaa704a-4fd9-4447-8963-9012da81ffc0'.</p>

**Comment by damirarh** [comment]
Imported comment written by **Damir Arh** on **2015-09-06T19:36:56**

<p>I assume you open the application again within a short period of time. This probably means that the instance owner from before you closed the application is still holding the lock, because HostLockRenewalPeriod has not passed yet.</p><p>When you don't need the instance store any more before you close the application, you should delete the instance owner:</p><p>var deleteOwnerCmd = new DeleteWorkflowOwnerCommand();<br>instanceStore.Execute(instanceHandle, deleteOwnerCmd, TimeSpan.FromSeconds(10));</p><p>This should release the lock immediately, without waiting for the HostLockRenewalPeriod to pass.</p>

**Comment by damirarh** [comment]
Imported comment written by **hamid rezaeirad** on **2015-09-07T07:02:51**

<p>Thank you for your reply.Yes, You right.I open the application for short period of time.I decrease the period of time and wait a few  minutes to open the application then it works.</p><p>I want to create SOA(service oriented architect) application with WCF and WF.I searched to find reference for create WF service as WCF.but I could n`t find any.could you please help me how to do that?</p>

**Comment by damirarh** [comment]
Imported comment written by **Damir Arh** on **2015-09-07T20:24:10**

<p>The best resource I've found on the topic is the book "Pro WF: Windows Workflow in .NET 4": <a href="http://amzn.to/1JNAgLM" rel="nofollow noopener" title="http://amzn.to/1JNAgLM">http://amzn.to/1JNAgLM</a>. It's great learning material for both Workflow Services (that's WF based WCF services you're interested in), and Workflow Foundation 4 in general. Note, that there's also a newer book "Pro WF 4.5" from the same publisher, but it's written by a different author: <a href="http://amzn.to/1XCwY7g" rel="nofollow noopener" title="http://amzn.to/1XCwY7g">http://amzn.to/1XCwY7g</a>. I haven't read that one, so I can't say how good it is.</p><p>There's not a lot of online resources on the topic, though. To get you started, you could take a look at Workflow Services on MSDN: <a href="https://msdn.microsoft.com/en-us/library/dd456788.aspx" rel="nofollow noopener" title="https://msdn.microsoft.com/en-us/library/dd456788.aspx">https://msdn.microsoft.com/...</a>.</p>

**Comment by damirarh** [comment]
Imported comment written by **Nayana Nayak** on **2016-02-01T08:31:38**

<p>I want to persist the workflow instance and load it.I create a WF app and reference it to a windows application and persist the WF and it is working fine with respect to windows application. but when I create same with reference wit the web applications its showing following errors :</p><p>The execution of an InstancePersistenceCommand was interrupted because the instance 'ba8ab62d-5642-4ebe-9c9d-bbb9d41e3dbc' is locked by a different instance owner. This error usually occurs because a different host has the instance loaded. The instance owner ID of the owner or host with a lock on the instance is 'ccaa704a-4fd9-4447-8963-9012da81ffc0'</p>

**Comment by damirarh** [comment]
Imported comment written by **Damir Arh** on **2016-02-01T17:53:29**

<p>It's a similar issue to the one, discussed with Hamid. Since this is  happening to you only in web applications, you are probably attempting to access the same instance from multiple requests in parallel or in short succession, i.e. before the HostLockRenewalPeriod has passed.</p><p>You'll need to avoid doing that. I suggest you manage access to workflows at the web application level, not the request level. You'll need to have only a single (global) instance store for accessing a single workflow instance, which will allow you to serialize access to it from different requests - the next request will only be given access to it, after the previous one completes. By using the same instance store for all requests you will also avoid any issues caused by the HostLockRenewalPeriod being to long.</p><p>You might even be able to have a single instance store for all requests, no matter the instance they are accessing. Make sure you do some performance tests to make sure this doesn't cause bottlenecks.</p><p>Web farm scenarios will make this much more difficult. If you have such requirements, it would be best to balance the load based on the workflow instance a request needs to access.</p>

**Comment by damirarh** [comment]
Imported comment written by **Parag** on **2016-02-18T06:22:41**

<p>Hi Damir,<br>When we have many workflows is it still better to have single store (of course given performance is fine) or we should keep different instance store for each type of workflow?</p><p>Another question regarding resuming runnable instance when using delay activity which I couldn't find answer yet. We are using 4.5, have a long running workflow in which have a pick activity with bookmark and delay activity being triggers. If bookmark is not called and delay expires workflow should become runnable and execute next steps. To start the workflow we are using identity, to pass in WorkflowApplication (e.g. WorkflowApplication instance = new WorkflowApplication(wf, inputs, IdentityV1);). Workflow host then waits for events as suggested in many places -</p><p>private static void WaitForRunnableInstance(InstanceStore store, InstanceHandle ownerHandle)<br>        {<br>            IEnumerable&lt;instancepersistenceevent&gt; events = store.WaitForEvents(ownerHandle, TimeSpan.MaxValue);<br>            bool foundRunnable = false;<br>            // Loop through the persistence events looking for the HasRunnableWorkflow event (in this sample, it corresponds with</p><p>            // the workflow instance whose timer has expired and is ready to be resumed by the host).</p><p>            foreach (InstancePersistenceEvent persistenceEvent in events)<br>            {<br>                if (persistenceEvent.Equals(HasRunnableWorkflowEvent.Value))<br>                {<br>                    foundRunnable = true;<br>                    Console.WriteLine("Found runnable instance");<br>                    break;<br>                }<br>            }<br>            if (!foundRunnable)<br>            {<br>                throw new ApplicationException("Unexpected: No runnable instances found in the instance store");<br>            }<br>        }</p><p>But when we use identity waitForEvents never generate events and stuck forever in above method. If we dont use identity in WorkflowApplication it works. There was a solution (second last in thread) at this link <a href="https://social.msdn.microsoft.com/Forums/vstudio/en-US/8a424c0b-44f8-4670-8d64-9c7142117b55/instancestore-waitforevents-not-firing-when-using-workflowapplication-with-a-workflowidentity?forum=wfprerelease" rel="nofollow noopener" title="https://social.msdn.microsoft.com/Forums/vstudio/en-US/8a424c0b-44f8-4670-8d64-9c7142117b55/instancestore-waitforevents-not-firing-when-using-workflowapplication-with-a-workflowidentity?forum=wfprerelease">https://social.msdn.microso...</a> but this does not worked either.<br>Can you suggest solution to this issue? Or do we need to poll DB in our own thread to check runnable instances and run when found?</p>

**Comment by damirarh** [comment]
Imported comment written by **Damir Arh** on **2016-02-19T08:16:31**

<p>I don't think there's any reason to have different instance stores for different types of workflows, if you don't have any performance problems. In the product I was working on, we only had a single instance store for all the different workflow definitions, and never had any problems because of that.</p><p>Unfortunately, I don't have any suggestions for your second issue. I did some tests with dynamic update (which would require the use of workflow identity), but couldn't get it to work reliably enough to ever implement it in the product. Hence, I have very limited experience with workflow identity.</p>

**Comment by damirarh** [comment]
Imported comment written by **Parag** on **2016-02-19T08:33:39**

<p>Thanks Damir for reply, but that issue is resolved, I was using CreateWorkflowOwnerCommand instead of CreateWorkflowOwnerWithIdentityCommand as suggested on msdn post.<br>Another issue I am trying to find answer now - In your workflows did you had delay activities, if yes, then how did you handled loading runnable instances when using single store?</p>

**Comment by damirarh** [comment]
Imported comment written by **Damir Arh** on **2016-02-19T11:10:16**

<p>No, we didn't have a need for delay activities, nor did we support them. We did store additional data about each instance in the database to know the workflow definition it mapped to, if that's the problem you are encountering.</p>

**Comment by damirarh** [comment]
Imported comment written by **Parag** on **2016-03-01T14:28:47**

<p>Hi Damir, thank you for the pointers, I used "workflow definition" - "workflow host" relation to reload the runnable workflow when delay is expired and its working fine as of now.</p>

**Comment by damirarh** [comment]
Imported comment written by **Sergey 222** on **2016-03-14T10:44:14**

<p>Hello guys!<br>I've got the same problem with absence of events in store while using wf identity. Parag, could you please help me with this issue. I think my code is good, but seems I've lost some minor detail...</p>

**Comment by damirarh** [comment]
Imported comment written by **Parag** on **2016-03-14T12:39:40**

<p>Hi Sergey,<br>In my case issue was with creating handle, I was using CreateWorkflowOwnerCommand instead of CreateWorkflowOwnerWithIdentityCommand. Now my code to create handle is as below -<br>Please note the xnamespaces, specially WorkflowHostTypePropertyName defined in my method below should use 4.0 properties not 4.5.<br>I hope this helps.</p><p></p><pre>private static readonly XNamespace _workflowNamespace = XNamespace.Get("urn:schemas-microsoft-com:System.Activities/4.0/properties");<br><br>        private static readonly XNamespace _workflow45Namespace = XNamespace.Get("urn:schemas-microsoft-com:System.Activities/4.5/properties");<br><br>private void CreateHandle()<br><br>        {<br><br>            _handle = _instanceStore.CreateInstanceHandle();<br><br>            XName WorkflowHostTypePropertyName = _workflowNamespace.GetName("WorkflowHostType");<br><br>            XName DefinitionIdentityFilterName = _workflow45Namespace.GetName("DefinitionIdentityFilter");<br><br>            XName WorkflowApplicationName = _workflow45Namespace.GetName("WorkflowApplication");<br><br>            XName DefinitionIdentitiesName = _workflow45Namespace.GetName("DefinitionIdentities");<br><br>            CreateWorkflowOwnerWithIdentityCommand ownerCommand = new CreateWorkflowOwnerWithIdentityCommand()<br><br>            {<br><br>                InstanceOwnerMetadata =<br><br>                {<br><br>                    { WorkflowHostTypePropertyName, new InstanceValue(WorkflowApplicationName) },<br><br>                    { DefinitionIdentityFilterName, new InstanceValue(WorkflowIdentityFilter.Any) },<br><br>                }<br><br>            };<br><br>            _instanceStore.DefaultInstanceOwner = _instanceStore.Execute(_handle, ownerCommand, TimeSpan.MaxValue).InstanceOwner;<br><br>        }<br></pre>

**Comment by damirarh** [comment]
Imported comment written by **Sergey 222** on **2016-03-14T14:59:06**

<p>Ahahah, it works! Thank you!</p>
